recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,266 @@
+| -------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| OCI compliant registry | An OCI-compliant registry refers to a container registry that adheres to the specifications set by the Open Container Initiative  |
+| ORAS | ORAS is the de facto tool for working with OCI Artifacts.(https://oras.land/) |
+| Private Bicep Registry | A private Bicep registry is a dedicated repository for storing and managing Bicep modules which is not accessible publicly need to authenticate into it for registry operations.| 
+
+## Objectives
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6917
+
+### Goals
+- Enable support to use bicep recipes stored in private bicep registries.
+- Support all OCI compliant private registries.
+
+### Non goals
+
+- Federated authentication(Azure workload identity/ AWS IRSA) for ACR and ECR.(will be part of initial feature release, its in investigation phase) 
+- Support to manage(configure/view) OCI registry credentials in Radius environment via CLI. (this is a future priority scenario and would not be in scope for the initial release)
+
+### User scenarios (optional)
+As an operator I am responsible for maintaining Bicep recipes for use with Radius. Bicep recipes used contains proprietary configurations, or data that should not be shared publicly and its intended for internal use within our organization and have granular control over who can access, contribute to, or modify bicep recipes. So I store the recipes in a private bicep registry. And I would like to use these private registries as template-path while registering a bicep recipe.
+
+
+## User Experience (if applicable)
+Credential information is stored in `Application.Core/secretstore` resource, users are expected to provide `username` and `password` keys with the appropriate values in the secret store that will be used to authenticate private registry.
+
+```
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'acrsecrets'
+  properties:{
+    type: 'generic'
+    resource: 'registry-secrets/acr'
+    data: {
+      'username': {
+        value: '<username>'
+      }
+      'password': {
+        value: '<registry-password>'
+      }
+    }
+  }
+}
+```
+`bicep` recipe config is added in the environment to stores bicep recipe configurations, update it to point to the `secretStore` with credential information.
+```
+"recipeConfig": {
+  "terraform": {
+    ...
+  },
+  "bicep": {
+    "authentication":{
+      "test.azurecr.io":{
+        "secret": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/secretStores/acrsecrets"
```

Comment:
Will we need to make changes to this when we add AZWI support?

Reply:
How will our code tell the difference between AZWI and username+password?

Reply:
Updated the design to handle different registry auth types.